src/Symfony/Component/VarExporter/Instantiator.php

```diff
@@ -0,0 +1,104 @@
+     *     Instantiator::instantiate([
+     *         Foo::class => [],
+     *         Bar::class => ['privatePropertyName' => $propertyValue],
+     *     ]);
+     *
+     * Instances of ArrayObject, ArrayIterator and SplObjectHash can be created
+     * by using the special "\0" property name to define their internal value:
+     *
+     *     // creates an SplObjectHash where $info1 is attached to $obj1, etc.
+     *     Instantiator::instantiate([SplObjectStorage::class => ["\0" => [$obj1, $info1, $obj2, $info2...]]]);
+     *
+     *     // creates an ArrayObject populated with $inputArray
+     *     Instantiator::instantiate([ArrayObject::class => ["\0" => $inputArray]]);
+     *
+     * @param array $propertiesByClass The properties to set on the object, keyed by the class to bind to
+     *                                 when setting them. The first key of the array defines the class
+     *                                 of the returned instance.
+     *
+     * @return object An instance of the class defined by the first key of $propertiesByClass
+     *
+     * @throws \ReflectionException When the instance cannot be created
+     */
+    public static function instantiate(array $propertiesByClass)
+    {
+        if (!$propertiesByClass) {
+            throw new \ReflectionException('No class provided.');
+        }
+
+        foreach ($propertiesByClass as $class => $properties) {
+            break;
+        }
+
+        try {
+            $reflector = Registry::$reflectors[$class] ?? Registry::getClassReflector($class);
+        } catch (\Exception $e) {
+            if (0 === strpos($e->getMessage(), 'Serialization of ')) {
+                $e = new \ReflectionException(substr_replace($e->getMessage(), 'Instanti', 0, 8));
+            }
+
+            throw $e;
+        }
+
+        if (Registry::$cloneable[$class]) {
+            $instance = clone Registry::$prototypes[$class];
+        } elseif (Registry::$instantiableWithoutConstructor[$class]) {
+            $instance = $reflector->newInstanceWithoutConstructor();
+        } elseif (null === Registry::$prototypes[$class]) {
+            throw new \ReflectionException(sprintf("Instantiation of '%s' is not allowed", $class));
+        } elseif ($reflector->implementsInterface('Serializable')) {
+            $instance = unserialize('C:'.\strlen($class).':"'.$class.'":0:{}');
+        } else {
+            $instance = unserialize('O:'.\strlen($class).':"'.$class.'":0:{}');
+        }
+
+        $instance = array($instance);
+
+        foreach ($propertiesByClass as $class => $properties) {
+            if (!$properties) {
+                continue;
+            }
+            foreach ($properties as $name => $value) {
+                $properties[$name] = array($value);
+            }
+            (Hydrator::$hydrators[$class] ?? Hydrator::getHydrator($class))($properties, $instance);
```

Comment:
I wrote a little component called `ocramius/lazy-map` for these cases: unsure if worth using its approach here, as it may both speed up or slow down everything (needs benchmarks), but at least it gets rid of the null coalesce everywhere

Reply:
Thanks for the link, interesting. I this specific case, we should stay with `??` as it is really the best way when using arrays directly (no deps+perf-wise).